haystack/nodes/other/join_docs.py

```diff
@@ -41,33 +48,38 @@ def __init__(
         self.top_k_join = top_k_join
 
     def run(self, inputs: List[dict], top_k_join: Optional[int] = None):  # type: ignore
-        if self.join_mode == "concatenate":
-            document_map = {}
-            for input_from_node in inputs:
-                for doc in input_from_node["documents"]:
-                    document_map[doc.id] = doc
-        elif self.join_mode == "merge":
-            document_map = {}
-            if self.weights:
-                weights = self.weights
-            else:
-                weights = [1 / len(inputs)] * len(inputs)
-            for input_from_node, weight in zip(inputs, weights):
-                for doc in input_from_node["documents"]:
-                    if document_map.get(doc.id):  # document already exists; update score
-                        document_map[doc.id].score += doc.score * weight
-                    else:  # add the document in map
-                        document_map[doc.id] = deepcopy(doc)
-                        document_map[doc.id].score *= weight
-        else:
-            raise Exception(f"Invalid join_mode: {self.join_mode}")
-
-        documents = sorted(document_map.values(), key=lambda d: d.score, reverse=True)
-
-        if top_k_join is None:
-            top_k_join = self.top_k_join
+        results = [inp["documents"] for inp in inputs]
+        weights = self.weights if self.weights else [1 / len(inputs)] * len(inputs)
+        document_map = {doc.id: [doc, 0] for doc in itertools.chain(*results)}
```

Comment:
Two notes here:

1. Why `for doc in itertools.chain(*results)}`? `results` is iterable. I think I'm missing something here.

2. `{doc.id: [doc, 0]}` looks like a more complex data structure than it's needed. It simply makes accessing the proper values harder. I would recommend splitting it into `documents_map` and `scores_map`.

That would make this line become two lines:
```python
document_map = {doc.id: doc for doc in results}
scores_map = {doc.id: 0 for doc in results}
```
In this way:

- Line 58 is simplified (`scores_map[doc.id] = doc.score`), as well as lines 60 and 62
- Line 66 is slightly changed: `sorted_scores = sorted(scores_map.items(), key=lambda d: d[1], reverse=True)`
- Block 67-71 becomes a bit clearer:
```python
docs = []
for idx in range(top_k_join):
    doc_id, score = sorted_scores[idx]
    doc = documents_map[doc_id]
    doc.score = score
    docs.append(doc)
```
Note that you don't need to iterate on `sorted_scores`: you just need the topK docs, so rather stop there :slightly_smiling_face: